src/cljam/util/sam_util.clj

```diff
@@ -11,29 +11,49 @@
 (defn- parse-header-keyvalues
   "e.g. \"LN:45 SN:ref\" -> {:LN 45, :SN \"ref\"}"
   [keyvalues]
-  (apply merge
-         (map (fn [kv]
-                (let [[k v] (cstr/split kv #":")]
-                  {(keyword k) (case k
-                                 "LN" (Integer/parseInt v)
-                                 "PI" (Integer/parseInt v)
-                                 v)}))
-              keyvalues)))
-
-(defn parse-header-line [line]
+  (into
+   {}
+   (map (fn [kv]
+          (let [[k v] (cstr/split kv #":")]
+            [(keyword k)
+             (case k
+               "LN" (Integer/parseInt v)
+               "PI" (Integer/parseInt v)
+               v)])))
+   keyvalues))
+
+(defn parse-header-line
+  "e.g. \"@SQ	SN:ref	LN:45\" => [:SQ {:SN \"ref\" :LN 45}]"
+  [line]
   (let [[typ & kvs] (cstr/split line #"\t")]
-    {(keyword (subs typ 1)) (if (= typ "@HD")
-                              (parse-header-keyvalues kvs)
-                              (vector (parse-header-keyvalues kvs)))}))
-
-(defn- parse-header* [col]
-  (when (seq col)
-    (merge-with #(vec (concat %1 %2)) (parse-header-line (first col)) (parse-header* (rest col)))))
+    [(keyword (subs typ 1)) (parse-header-keyvalues kvs)]))
+
+(defn finalize
+  "Wrap a reducing function with finalizing function."
+  [rf f]
+  (fn
+    ([] (rf))
+    ([x] (f (rf x)))
+    ([x y] (rf x y))))
+
+(defn into-with-conj
+  "A reducing function that creates a map of vector.
+   e.g. (transduce identity into-with-conj [[:a 1] [:b 2] [:a 3]]) => {:a [1 3] :b [2]}"
+  ([] (transient {}))
```

Comment:
`into-with-conj` and `finalize` are not so general. It can be only used in transducers. Non-argument arity of `into-with-conj`, especially, is not good for common use because it returns transient collection. I recommend making these functions private.